app/Http/Controllers/Legacy/Web/CampaignsController.php

```diff
@@ -18,56 +19,74 @@ public function __construct()
         $this->middleware('role:admin,staff', ['only' => ['store', 'update', 'destroy']]);
     }
 
+    /**
+     * Show overview of campaigns and their IDs.
+     */
+    public function index()
+    {
+        $campaigns = Campaign::all();
+
+        // Format start and end dates how we want them to be viewed.
+        foreach ($campaigns as $campaign) {
+            $campaign = $this->formatStartEndDates($campaign, '');
+        }
+
+        return view('pages.campaigns_index')->with('campaigns', $campaigns);
+    }
+
     /**
      * Store a newly created resource in storage.
      *
      * @param  \Illuminate\Http\Request  $request
-     * @return \Illuminate\Http\Response
      */
     public function store(Request $request)
     {
         $this->validate($request, [
             'internal_title' => 'required|string|unique:campaigns',
             'start_date' => 'required|date',
-            'end_date' => 'required|date',
+            'end_date' => 'nullable|date|after:start_date',
         ]);
 
-        $campaign = Campaign::create($request->all());
+        // Change dates to YYYY-MM-DD format so it will save in the database.
+        $campaignDetails = $this->formatDatesForDatabase($request->all());
```

Comment:
This should be done with a [date mutator](https://laravel.com/docs/5.7/eloquent-mutators#date-mutators) on the model.

Reply:
@DFurnes not sure what you mean by this - do you mean setting it in a different format to in `$dateFormat`?

Reply:
You don't need to manually iterate over the request and convert to `Carbon` instances here. If you set the `$dates` property on the `Campaign` model to include `start_date` and `end_date`, this will be handled automatically. In `Campaign.php`:

```php
class Campaign
{
    /**
     * The attributes that should be mutated to dates.
     *
     * @var array
     */
    protected $dates = [
        'created_at',
        'updated_at',
        'start_date',
        'end_date',
    ];
```

That will remove the need for the `formatDatesForDatabase` method in this controller, and the need for the `formatStartEndDates` method (by allowing you to directly call `format()` on the Carbon instances for those properties).

Reply:
@DFurnes thanks for explaining more! That is what I thought you meant but when I change that and just send `$request->all()` to `Campaign::create($`request->all()`, it gives me this error: 
![screen shot 2018-11-06 at 5 13 29 pm](https://user-images.githubusercontent.com/9019452/48097270-669c1480-e1e7-11e8-9132-b207e76b4983.png)

Reply:
Oh, right! Frustratingly, Laravel supports auto-serializing a couple of formats (like UNIX timestamps and `YYYY-MM-DD`), but not `MM/DD/YYYY`. We built an ["arbitrary date" mutator](https://github.com/DoSomething/northstar/blob/3e37877184882d280f3574ebd09893f6a931c32e/app/Models/User.php#L223-L252) for this purpose in Northstar that'd make sense to snag here.

Reply:
@chloealee You can handle this with a mutator, like we do for the `last_messaged_at` field in Northstar in the [code sample](
https://github.com/DoSomething/northstar/blob/3e37877184882d280f3574ebd09893f6a931c32e/app/Models/User.php#L223-L252) I linked above! So copy over that helper method to the model, and then you can use it for each field. For example, for `start_date`:

```php
    /**
     * Mutator for setting the start_date field.
     *
     * @param string|Carbon $value
     */
    public function setStartDateAttribute($value)
    {
        $this->setArbitraryDateString('start_date', $value);
    }
```

Reply:
sry @DFurnes I'm still confused how this works! How does `setStartDateAttribute` get called to mutate the date? Is that just some Laravel magic?

Reply:
(but it is updated! and works)

Reply:
Yep, that's an [Eloquent mutator](https://laravel.com/docs/5.7/eloquent-mutators#defining-a-mutator)! When setting a field, Laravel will check whether there's a mutator with a matching name before saving the variable.

Reply:
very cool - that is exactly what I was looking for in the docs - thanks so much @DFurnes !